---
title: "Quiz3"
author: "Syeda"
date: "4/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Question 1
Question: What is the mean expression across all features for sample 5 in the ALL dataset (from the ALL package)?
```{r, echo=FALSE}
#Install biomaRt
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")


```


```{r 1, echo=FALSE}
# Install ALL library
library(ALL)
#Get expression data for 5th column of ALL dataset
expr_data5 <- exprs(ALL)[,5]
mean(expr_data5)
```
Expected answer 5.629627

Question 2.
We will use the biomaRt package to annotate an Affymetrix microarray. We want our results in the hg19 build of the human genome and we therefore need to connect to Ensembl 75 which is the latest release on this genome version. How to connect to older versions of Ensembl is described in the biomaRt package vignette; it can be achived with the command 

mart <- useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")

Question: Using this version of Ensembl, annotate each feature of the ALL dataset with the Ensembl gene id. How many probesets (features) are annotated with more than one Ensembl gene id?

1 point

```{r , echo=FALSE}
library(biomaRt)
# Sse the archived versions
listEnsemblArchives()
#To download and select Ensembl 75 as our mart/database
mart <- useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")
mart
#List datasets inside the database/mart
head(listDatasets(mart))
# Choose the human dataset
ensembl <- useDataset("hsapiens_gene_ensembl", mart)
ensembl

# Get the Affymetrix probe ID's
ids <- featureNames(ALL)
head(ids)
length(ids)

#Map ids to gene names
#Look for the Annotation Package used in ALL
#Choose appropriate filter
listFilters(ensembl)
#
gene_probes <- getBM(attributes = c("ensembl_gene_id","affy_hg_u95av2"), filters = "affy_hg_u95av2", values = ids, mart = ensembl)
   
library(dplyr)
  

#Sum the number of probesets that have more than one ensemble gene id associated with it
sum(tbl$num_probes)
```
Expected answer : 1045

```{r 3, echo=FALSE}

library(biomaRt)
# Sse the archived versions
listEnsemblArchives()
#To download and select Ensembl 75 as our mart/database
mart <- useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")
mart
#List datasets inside the database/mart
head(listDatasets(mart))
# Choose the human dataset
ensembl <- useDataset("hsapiens_gene_ensembl", mart)
ensembl

# Get the Affymetrix probe ID's
affy_ids <- featureNames(ALL)

gene_probes_chr <- getBM(attributes = c("ensembl_gene_id","affy_hg_u95av2","chromosome_name"), filters = "affy_hg_u95av2", values = affy_ids, mart = ensembl)

```

```{r 3, echo=FALSE}

# Group table by affymetrix probe ids
group_tbl <- gene_probes_chr %>% group_by(affy_hg_u95av2) 
#Create range for autosomes
range <- (1:22)
# Select only rows that belong to range
select_ids <- with(tmp_tbl, affy_hg_u95av2[chromosome_name %in% range])
# Choose only unique ids and sum
length(unique(select_ids))
```

Expected answer: 11016


Question 4
Use the MsetEx dataset from the minfiData package. Part of this question is to use the help system to figure out how to address the question.

Question: What is the mean value of the Methylation channel across the features for sample “5723646052_R04C01”?



```{r 4,echo=FALSE}
#Install minfi package

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")

#Load package
library(minfi)

# Install minfiData package

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("minfiData")
#Load package
library(minfiData)
class?MethylSet
#Get Methylation matrix
meth_data <- getmeth(MsetEx)
# Get mean of sample “5723646052_R04C01”
mean(meth_data[,2])

```
Expected answer: 7228.277


Question 5
Question: Access the processed data from NCBI GEO Accession number GSE788. What is the mean expression level of sample GSM9024?

```{r 5, echo=FALSE}
#Load GEOquery package
library(GEOquery)
#Get data for NCBI GEO Accession number GSE788
eList <- getGEO("GSE788")
length(eList)
# To see the name of the elament here series matrix with accession number
names(eList)
#Get data
eData <- eList[[1]]

#Get Expression matrix for eData
exp_data <- exprs(eData)
#Get the mean for sample col GSM9024
mean(exp_data[,2])

```
Expected answer: 756.432

Question 6
We are using the airway dataset from the airway package.

Question: What is the average of the average length across the samples in the expriment?

```{r 6, echo=FALSE}

#Example data package
library(airway)
#Load data
data(airway)
# Get the colData(phenotype Data)
col_data <- colData(airway)
mean(col_data$avgLength)
```
Expected answer: 113.75


```{r 7, echo=FALSE}

# Get assay accesor 
airway

# Get expression data
expr_data <- assay(airway, "counts")
sample1 <- expr_data[,3]
#The number of genes which have a count of 1 read or more in sample SRR1039512/col 3
sum(sample1>0)
```


```{r 8,echo= }

library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
#Create txdb object
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
#Get exons of txdb
txdb_exons <- exons(txdb)
#Cretae list of autosomes
chromosomes <- paste("chr", 1:22, sep="")
#Select exons for autosomes only
txdb_aut_gr <- keepSeqlevels(txdb_exons, chromosomes, pruning.mode = "coarse")

# Map transcript ranges to NCBI style
new_style <- mapSeqlevels(seqlevels(txdb_aut_gr), "NCBI")
#Rename seqnames of transcripts to NCBI style
tx_ncbi_gr <- renameSeqlevels(txdb_aut_gr, new_style)



#load GenomicRanges
library(GenomicRanges)
library(SummarizedExperiment)
# Get GRanges from airway
airway_ranges <- rowRanges(airway)

#Subset by overlaps to get overlap between GRangesList and Transcript ranges
olap_gr <- subsetByOverlaps(airway_ranges, tx_ncbi_gr)
# Compute the  number of features overlapping with transcripts on autosomes
length(olap_gr)


```

Expected answer: 26276


Question 9
The expression measures of the airway dataset are the number of reads mapping to each feature. In the previous question we have established that many of these features do not overlap autosomal transcripts from the TxDb.Hsapiens.UCSC.hg19.knownGene. But how many reads map to features which overlaps these transcripts?

Question: For sample SRR1039508, how big a percentage (expressed as a number between 0 and 1) of the total reads in the airway dataset for that sample, are part of a feature which overlaps an autosomal TxDb.Hsapiens.UCSC.hg19.knownGene transcript?

```{r 9, echo=FALSE}
library(airway)
#Get expression matrix
# Get assay accesor 
airway
# Get expression data
expr_data <- assay(airway, "counts")
#Get reads for sample SRR1039508 
reads_sample1 <- expr_data[,1]
#Sum of all reads for sample1 SRR1039508
sum_sample1_reads <- sum(reads_sample1)


# Get the rownames for all rows of genes in overlapping region
r_names <- elementNROWS(olap_gr)
# Get names of overlapping genes
olap_genes <- names(r_names)
# Subset expression data to get only features with genes overlapping transcripts

expr_data_olap <- subset(expr_data, rownames(expr_data) %in% olap_genes)
#Get features that overlap from sample SRR1039508
reads_olap <- expr_data_olap[,1]
sum_olap_reads <- sum(reads_olap)

#Caluculate percent that overlap for sample1
sum_olap_reads/sum_sample1_reads

```
Expected answer: 0.9004193


Question 10
Consider sample SRR1039508 and only consider features which overlap autosomal transcripts from TxDb.Hsapiens.UCSC.hg19.knownGene. We should be able to very roughly divide these transcripts into expressed and non expressed transcript. Expressed transcripts should be marked by H3K4me3 at their promoter. The airway dataset have assayed “airway smooth muscle cells”. In the Roadmap Epigenomics data set, the E096 is supposed to be “lung”. Obtain the H3K4me3 narrowPeaks from the E096 sample using the AnnotationHub package.

Question: What is the median number of counts per feature (for sample SRR1039508) containing a H3K4me narrowPeak in their promoter (only features which overlap autosomal transcripts from TxDb.Hsapiens.UCSC.hg19.knownGene are considered)?

Clarification: We are using the standard 2.2kb default Bioconductor promoter setting.

Conclusion Compare this to the median number of counts for features without a H3K4me3 peak. Note that this short analysis has not taken transcript lengths into account and it compares different genomic regions to each other; this is highly susceptible to bias such as sequence bias.

```{r 10,echo=FALSE}

#Get txdb exons
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
#Create txdb object
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
#Get exons of txdb
txdb_exons <- exons(txdb)


# Get H3K4me3 peaks
library(AnnotationHub)
ah <- AnnotationHub()
ah <- subset(ah, species == "Homo sapiens")
qh <- query(ah, c("H3K4me3","E096"))
#Get the narrowPeak data for E096 sample
h3k_data <- qh[["AH30596"]]

#Overlap transcripts with peaks
tx_h3k_olap <- subsetByOverlaps(txdb_exons,h3k_data)
#Keep only the Ranges for autosomes
#Cretae list of autosomes
chromosomes <- paste("chr", 1:22, sep="")
#Select transcripts/exons for autosomes only
tx_h3k_aut <- keepSeqlevels(tx_h3k_olap, chromosomes, pruning.mode = "coarse")
# Map transcript ranges to NCBI style
new_style <- mapSeqlevels(seqlevels(tx_h3k_aut), "NCBI")
#Rename seqnames of transcripts to NCBI style
tx_h3k_ncbi_gr <- renameSeqlevels(tx_h3k_aut, new_style)

#Get features from sample1 in airways that overlap these transcripts with peaks
tx_h3k_airway_olap <- subsetByOverlaps(airway_ranges, tx_h3k_ncbi_gr)
# Get promoters for the transcripts with peaks
prom_peak_gr <- promoters(tx_h3k_airway_olap)
#Unlist ot ro get GRanes
unlisted_prompeak_gr <- unlist(prom_peak_gr)
#Get names of features from the above
# Get the rownames for all rows of genes in overlapping region
row_names <- elementNROWS(unlisted_prompeak_gr)
# Get names of overlapping genes
genes_peaks <- names(row_names)
# Subset expression data to get only features with genes overlapping transcripts

#Subset the expression data with the above features
expr_peak_olap <- subset(expr_data, rownames(expr_data) %in% genes_peaks)
#Get the sample1 column for these features
olap_features <- expr_peak_olap[,1]

median(expr_peak_olap[,1])  
#This gives 234 as answer 
```

Expected Answer: 232




