{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Final_Python_for_genomics.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyPbQg+OfMR9xFnRylTQ2lO3"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"FMOTjBpj8Nm5"},"source":["Final Exam 1 - 3"]},{"cell_type":"code","metadata":{"id":"iLZ9iiXy9PNg"},"source":["# Mount google drive\r\n","\r\n","from google.colab import drive\r\n","drive.mount('/content/gdrive')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"LaMZ4J9m9MVY"},"source":["#Change working directory to Genomic Data Science\r\n","%cd /content/gdrive/My\\ Drive/'Genomic Data Science'"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"5WUtzzjL9KMQ"},"source":["#Check if file exists\r\n","\r\n","try:\r\n","  f = open('dna2.fasta') # Read file \r\n","except IOError:\r\n","  print(\"File does not exist!\") # Error message if file does not exist  "],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"phTgmD249C7x"},"source":["# Calculate the numbr of records in FASTA file\r\n","count = 0\r\n","seqs = {} # Empty dictionary \r\n","for line in f:\r\n","  line = line.rstrip() #Discards new line at end if any\r\n","  if line[0] == \">\":\r\n","    count += 1  \r\n","    words = line.split() # Split the header into words at spaces\r\n","    name = words[0][1:] # Get the name which is the first element of the list\r\n","    seqs[name] = \"\" # If its a header it inititalizes the value to empty string\r\n","\r\n","  else:\r\n","    seqs[name] = seqs[name] + line # If not header then add line as value to the key name       \r\n","print(\"Number of records:\",count)\r\n","print(len(seqs))\r\n","f.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"e6dByYVM9UXw"},"source":["Number of records = 18"]},{"cell_type":"code","metadata":{"id":"Opucxw2K8TbJ"},"source":["# Get the sequence with the max length\r\n","max_len = 0\r\n","max_id = \"\"\r\n","for record in SeqIO.parse(\"dna2.fasta\", \"fasta\"):\r\n","  if len(record) > max_len:\r\n","    max_len = len(record)\r\n","    seq_id = record.id\r\n","\r\n","\r\n","print(max_len)    \r\n","print(seq_id)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pN-PQnWY8X6J"},"source":["max length = 4894\r\n","gi|142022655|gb|EQ086233.1|255"]},{"cell_type":"code","metadata":{"id":"hdpomjGz8W5R"},"source":["#Get sequence with min length\r\n","min_len = 10**10\r\n","for record in SeqIO.parse(\"dna2.fasta\", \"fasta\"):\r\n","  if len(record) < min_len:\r\n","    min_len = len(record)\r\n","    seq_id = record.id\r\n","\r\n","print(min_len)    \r\n","print(seq_id)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"YgRmfn2q8fdR"},"source":["min length = 115\r\n","gi|142022655|gb|EQ086233.1|346"]},{"cell_type":"code","metadata":{"id":"8XMWmA1y5pW1"},"source":["# Mount google drive\r\n","from google.colab import drive\r\n","drive.mount('/content/gdrive')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"iUyKC-S754NC"},"source":["#Change working directory to Genomic Data Science\r\n","%cd /content/gdrive/My\\ Drive/'Genomic Data Science'"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"3koF-k17582K"},"source":["#Check if file exists\r\n","\r\n","try:\r\n","  f = open('dna2.fasta') # Read file \r\n","except IOError:\r\n","  print(\"File does not exist!\") # Error message if file does not exist  "],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"3MxYVOoQ5-K5"},"source":["#Install Biopython\r\n","!pip install Biopython"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"lNK94MfC6E1R"},"source":["# Import SeqIO module\r\n","import Bio\r\n","from Bio import SeqIO # Module for handling Sequence Input and Output operations"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"K0jDlGK16J-S"},"source":["# Print id's and length of the sequence in the record\r\n","for record in SeqIO.parse('dna2.fasta','fasta'):\r\n","  print(record.id, len(record)) # Prints record id's and sequence length"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Q7uQb1KZ6u4h"},"source":["Final Exam 4 - 7"]},{"cell_type":"code","metadata":{"id":"QTbLx9fn6Mjy"},"source":["# Functions to find start and stop codons in sequence(FINAL EXAM)\r\n","def has_start_codon(dna, frame):\r\n","  start_codon_found = False\r\n","  start = -1\r\n","  for i in range(frame, len(dna),3):\r\n","    codon = dna[i:i+3].lower()\r\n","    if codon == \"atg\":\r\n","      start_codon_found = True\r\n","      start = i\r\n","      break\r\n","  #print(start_codon_found)    \r\n","  return start_codon_found, start   \r\n","\r\n","#Has stop codon\r\n","def has_stop_codon(dna, frame):\r\n","  stop_codon_found = False\r\n","  stop = -1\r\n","  stop_codons = ['tga', 'tag', 'taa']\r\n","  for i in range(frame, len(dna), 3):\r\n","    codon = dna[i:i+3].lower()\r\n","    if codon in stop_codons:\r\n","      stop_codon_found = True\r\n","      stop = i\r\n","      break\r\n","\r\n","  #print(stop_codon_found)    \r\n","  return stop_codon_found, stop    "],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_LPesH8s6XcS"},"source":["# Checks for all start and stop codons and gives the longest orf in each frame(FINAL EXAM)\r\n","\r\n","orf_dict = {}\r\n","for record in  SeqIO.parse(\"dna2.fasta\", \"fasta\"):\r\n","  sequence, id = record.seq, record.id\r\n","  for position in range(len(sequence)):\r\n","    start_pos = position\r\n","    start1_found, start1 = has_start_codon(sequence,start_pos)\r\n","    if start1_found:\r\n","      #print(\"start:\", start1)\r\n","      stop1_found, stop1 = has_stop_codon(sequence,start1)\r\n","      if stop1_found and start1 < stop1:\r\n","        #print(\"stop:\", stop1)\r\n","        #print(\"sequence id:\", id)\r\n","        orf = sequence[start1:stop1] \r\n","        #print(\"length of ORF:\", len(orf))\r\n","        frame = start1%3\r\n","        #print(\"frame:\", start1%3)\r\n","        if frame not in orf_dict:\r\n","          orf_dict[frame]= (len(orf), start1, stop1, id)\r\n","        if len(orf) > orf_dict[frame][0]:\r\n","          orf_dict[frame] = (len(orf), start1, stop1, id)\r\n","print(orf_dict)  "],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"yA6r2jjs6f_n"},"source":["**expected answers:**\r\n","\r\n","2391 + 3 = 2394 is longest orf in any of the three frames(because we have to include the 3 bases of the stop codon)\r\n","\r\n","1455 + 3 = 1458 is the longest orf in frame 2\r\n"]},{"cell_type":"code","metadata":{"id":"8abaX3Kt6kG5"},"source":["# Find the longest forward orf for sequence with id \"gi|142022655|gb|EQ086233.1|16\"\r\n","from Bio import SeqIO\r\n","\r\n","record_dict = SeqIO.to_dict(SeqIO.parse(\"dna2.fasta\", \"fasta\"))\r\n","my_record = record_dict[\"gi|142022655|gb|EQ086233.1|16\"] # use any record ID\r\n","my_record  \r\n","my_sequence = my_record.seq\r\n","len(my_sequence)\r\n","len_orf = -1\r\n","for position in range(len(my_sequence)):\r\n","  start_pos = position\r\n","  start_found, start = has_start_codon(my_sequence, start_pos)\r\n","  if start_found:\r\n","    stop_found, stop = has_stop_codon(my_sequence, start)\r\n","    if start < stop:\r\n","      orf = my_sequence[start:stop]\r\n","      if len(orf) > len_orf:\r\n","        len_orf = len(orf)\r\n","print(len_orf)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"NCj0oEiW6pAZ"},"source":["expected answer for the above is 1641 + 3 = 1644 because we must include the 3 bases of the stop codon"]},{"cell_type":"markdown","metadata":{"id":"KBiSta_47551"},"source":["Final Exam 8 - 10"]},{"cell_type":"code","metadata":{"id":"5A0KisnM737J"},"source":["# Find repeats of length 12 that occur max times\r\n","# Function to count words of given length\r\n","#words=[]\r\n","#counts ={}\r\n","def split_into_words(dna,num):\r\n","  words=[]\r\n","  for i in range(len(dna)-num+1):\r\n","    word = dna[i:i+num]\r\n","    words.append(word)\r\n","  return words  \r\n","#print(words)  "],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"8EHjjMehztmx"},"source":["# Find most frequent repeat of length 6 in all sequences\r\n","records = list(SeqIO.parse(\"dna2.fasta\",\"fasta\"))\r\n","print(len(records))\r\n","count_of_repeats ={}\r\n","\r\n","sum1 = 0\r\n","for i in range(len(records)):\r\n","  my_seq = records[i].seq\r\n","  #print(my_seq)\r\n","  words_list = split_into_words(my_seq, 6)\r\n","  sum1 += len(words_list)\r\n","\r\n","  #print(words_list)\r\n","  for word in words_list:  \r\n","    if word not in count_of_repeats:\r\n","      count_of_repeats[word] = 1\r\n","    else:\r\n","      count_of_repeats[word] += 1\r\n","\r\n","#print(count_of_repeats)   \r\n","print(sorted(count_of_repeats.items(), reverse = True, key = lambda x: x[1]))\r\n","print(sum1, sum(count_of_repeats.values()))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4xKMLE0mzyjR"},"source":["\r\n","153 times GCGCGC"]},{"cell_type":"code","metadata":{"id":"JrYL0fwg0VCA"},"source":["# Find most frequent repeat of length 12 in all sequences\r\n","records = list(SeqIO.parse(\"dna2.fasta\",\"fasta\"))\r\n","print(len(records))\r\n","count_of_repeats ={}\r\n","\r\n","sum1 = 0\r\n","for i in range(len(records)):\r\n","  my_seq = records[i].seq\r\n","  #print(my_seq)\r\n","  words_list = split_into_words(my_seq, 12)\r\n","  sum1 += len(words_list)\r\n","\r\n","  #print(words_list)\r\n","  for word in words_list:  \r\n","    if word not in count_of_repeats:\r\n","      count_of_repeats[word] = 1\r\n","    else:\r\n","      count_of_repeats[word] += 1\r\n","\r\n","#print(count_of_repeats)   \r\n","print(sorted(count_of_repeats.items(), reverse = True, key = lambda x: x[1]))\r\n","print(sum1, sum(count_of_repeats.values()))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UvxSpixc0XRo"},"source":["Find all repeats of length 12 in the input file. Let's use Max to specify the number of copies\r\n","\r\n","of the most frequent repeat of length 12. How many different 12-base sequences\r\n","\r\n","occur Max times?\r\n","Three repeats of length 12 occur 4 times\r\n"]},{"cell_type":"code","metadata":{"id":"nBSFr-cS0Ww4"},"source":["# Find most frequent repeat of length 7 in all sequences\r\n","records = list(SeqIO.parse(\"dna2.fasta\",\"fasta\"))\r\n","print(len(records))\r\n","count_of_repeats ={}\r\n","\r\n","sum1 = 0\r\n","for i in range(len(records)):\r\n","  my_seq = records[i].seq\r\n","  #print(my_seq)\r\n","  words_list = split_into_words(my_seq, 7)\r\n","\r\n","  #print(words_list)\r\n","  for word in words_list:  \r\n","    if word not in count_of_repeats:\r\n","      count_of_repeats[word] = 1\r\n","    else:\r\n","      count_of_repeats[word] += 1\r\n","\r\n","#print(count_of_repeats)   \r\n","print(sorted(count_of_repeats.items(), reverse = True, key = lambda x: x[1]))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3Wluc34u08FP"},"source":["CGCGCCG occurs 63 times"]}]}